src/control/jobs.c

```diff
@@ -363,8 +406,9 @@ int dt_control_add_job(dt_control_t *control, dt_job_queue_t queue_id, _dt_job_t
 
         *queue = g_list_delete_link(*queue, iter);
         length--;
-        dt_control_job_set_state(job, DT_JOB_STATE_DISCARDED);
-        dt_control_job_dispose(job);
+
+        job_for_disposal = job;
```

Comment:
**job** is not in queue, so it shouldn't need locking access to it with queue_mutex.
Since job disposal will be slower than not job disposing, and there does not seem to be any reason to keep queue_mutex locked for disposing.

Since now i need to lock queue_mutex in dt_control_run_job() (to unset job from array of running jobs), this might slightly help, but i did not do any mutex performance measurements yet.